pkg/controller/route53hostedzone/zonereconciler.go

```diff
@@ -0,0 +1,336 @@
+		// IsNotFound is acceptable here. It just means that the object was likely deleted.
+		// We'll return a half constructed ZoneReconciler that has desiredState set to nil.
+		// That signals that a reconciliation does NOT need to happen since we don't know the
+		// desiredState.
+		zoneReconciler.logger.WithField("key", newZoneReconcilerInput.key).Debug("DNSZone has been deleted")
+		return zoneReconciler, nil
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	err = zoneReconciler.funcs.getCOAWSClient()
+	if err != nil {
+		return nil, err
+	}
+
+	err = zoneReconciler.funcs.getCurrentState()
+	if err != nil {
+		return nil, err
+	}
+
+	return zoneReconciler, nil
+}
+
+// setupFunctionPointers sets up the function pointers that the reconciler uses during reconciliation.
+// Overriding these function points is used to inject dependencies during unit testing.
+func (zr *ZoneReconciler) setupFunctionPointers(newZoneReconcilerInput *NewZoneReconcilerInput) {
+
+	// By default, we'll use these functions
+	zr.funcs = &functionPointers{
+		getDesiredState:         zr.getDesiredState,
+		getCOAWSClient:          zr.getCOAWSClient,
+		getCurrentState:         zr.getCurrentState,
+		createRoute53HostedZone: zr.createRoute53HostedZone,
+		deleteRoute53HostedZone: zr.deleteRoute53HostedZone,
+		awsClientBuilder:        clustopaws.NewClient,
+	}
+
+	if newZoneReconcilerInput.functionOverrides == nil {
+		// No overrides given, accept all defaults, nothing to do here.
+		return
+	}
+
+	// Setup any overrides for the default functions. This is useful for dependency injection.
+	if newZoneReconcilerInput.functionOverrides.getDesiredState != nil {
+		zr.funcs.getDesiredState = newZoneReconcilerInput.functionOverrides.getDesiredState
+	}
+
+	if newZoneReconcilerInput.functionOverrides.getCOAWSClient != nil {
+		zr.funcs.getCOAWSClient = newZoneReconcilerInput.functionOverrides.getCOAWSClient
+	}
+
+	if newZoneReconcilerInput.functionOverrides.getCurrentState != nil {
+		zr.funcs.getCurrentState = newZoneReconcilerInput.functionOverrides.getCurrentState
+	}
+
+	if newZoneReconcilerInput.functionOverrides.createRoute53HostedZone != nil {
+		zr.funcs.createRoute53HostedZone = newZoneReconcilerInput.functionOverrides.createRoute53HostedZone
+	}
+
+	if newZoneReconcilerInput.functionOverrides.deleteRoute53HostedZone != nil {
+		zr.funcs.deleteRoute53HostedZone = newZoneReconcilerInput.functionOverrides.deleteRoute53HostedZone
+	}
+
+	if newZoneReconcilerInput.functionOverrides.awsClientBuilder != nil {
+		zr.funcs.awsClientBuilder = newZoneReconcilerInput.functionOverrides.awsClientBuilder
+	}
+}
+
+// getDesiredState gets the current desired state from kubernetes.
+// If ZoneReconciler.desiredState is not nil, an error will be returned.
+func (zr *ZoneReconciler) getDesiredState() error {
+	if zr.desiredState != nil {
+		// This should never happen.
+		return fmt.Errorf("Desired State already populated for %v/%v", zr.namespace, zr.name)
+	}
```

Comment:
Per above I would pitch that the DNSZone be passed into the object directly by the controller and the tests.